### ISSUE TITLE
why my cls loss in training process always 0 

### ISSUE THREAD

**Issue by Ase21013** [eriklindernoren/PyTorch-YOLOv3#130]
I train this model for single class detection with my own dataset, others losses look ordinary and  downtrend, only cls loss always 0.
yolov3.cfg file i did modify(filter,class)

![1552815405558](https://user-images.githubusercontent.com/45848438/54488311-4ee24480-48db-11e9-96a8-a5a55855ae7d.gif)
this is part of my training process

i put my training weights in test.py  file to validate the model
After NMS
outputs are None
Does anyone meet this problem too?
and how to fix it

**Comment by fourth-archive** [comment]
@Ase21013 you could try this single-class training tutorial. It may be ok to see 0 cls loss.
https://github.com/ultralytics/yolov3/wiki/Example:-Train-Single-Class

**Comment by Ase21013** [comment]
@fourth-archive  
thank your reply 
I modify the file from this tutorial too. So it is normal phenomenon,
What cause the output from NMS is None?

**Comment by fourth-archive** [comment]
@Ase21013 use the repo in the tutorial, and repeat the steps exactly.

**Comment by Ase21013** [comment]
@fourth-archive 
thank your  reply 
it work!
thank you very much
![1234](https://user-images.githubusercontent.com/45848438/54598844-5ee15c00-4a74-11e9-8a85-536767a1798b.jpg)

**Comment by fourth-archive** [comment]
@Ase21013 those are impressive results, and an impressive hardware setup! Are you accessing those V100’s in the cloud? What dataset are you using? Is your test group separate from your train group?

**Comment by WannaSeaU** [comment]
@Ase21013  Hi,bro. I'am also training single class on my own dataset.
As I noticed, the 'cls' output is also always 0 in [this repo](https://github.com/ultralytics/yolov3/wiki/Example:-Train-Single-Class). 
So, is that normal? Did this training improve your model?

thanks.

**Comment by Ase21013** [comment]
@fourth-archive 
sorry, answer you so late.
Yes, my validation group separate from my train group
About V100's, those are our college's  server let researcher to use  for Deep learning. 
Training dataset is consisted of 512 driving recorder images, Validation is consisted of  256 images.

**Comment by Ase21013** [comment]
@WannaSeaU 
I thought this question's answer, this is my opinion.
If my opinion is wrong, please correct me.
Appreciate.
I read yolov3 paper and some information about  how to work and achieve before.
One of the processes of YOLOv3  will detect the part of image's area, 
output is background or a target.
In single class detection, if targets are not detected in area.
They will be classified into the background,  'cls'(class loss) Is calculated wrong classification,
background will not be included in the calculation and number of class is 1.
therefore target was  be detected,  it must be correct, so cls is 0.
That is what i thought.

**Comment by WannaSeaU** [comment]
@Ase21013 Thank you for your valuable insight. I may reply to you again if I get a different opinion.

**Comment by eriklindernoren** [comment]
I think when it comes to single class datasets since the classification part of the model is [only trained on areas where there exists objects](https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/master/models.py#L199) the model will quickly learn to always output `1`, since that will be the only ground truth value it encounters for the classification output. For this case the confidence output is more interesting, as that output will be trained to output `0` for background and `1` for objects (with the class you're interested in).

**Comment by LK-Peng** [comment]
> I think when it comes to single class datasets, since the classification part of the model is [only trained on areas where there are objects](https://github.com/eriklindernoren/PyTorch-YOLOv3/blob/master/models.py#L199), the model will quickly learn to always output `1`, since that will be the only ground truth value it encounters for the classification output.

Could you give some idea about how to trained on single class datasets? Thank you.

**Comment by Flova** [comment]
I closed this issue due to inactivity. Feel free to reopen for further discussion.
